Disable native appearance of devolvable widgets

obyknovenius opened the pull request:
#### 482584f17a16511d68223b83469e1dbf88085e2b
<pre>
Disable native appearance of devolvable widgets
<a href="https://bugs.webkit.org/show_bug.cgi?id=243899">https://bugs.webkit.org/show_bug.cgi?id=243899</a>
rdar://98899342

Reviewed by NOBODY (OOPS!).

All HTML widgets can be divided into two groups: devolvable and
non-devolvable. When certain CSS properties are applied to a devolvable
widget by author styles, the widget must be rendered in its primitive
appearance. Non-devolvable widgets remain native regardless of any such
properties.

Properties that disable native appearance of devolvable widgets are now
marked with `disables-native-appearance` key in `CSSProperites.json`.

<a href="https://html.spec.whatwg.org/multipage/rendering.html#widgets">https://html.spec.whatwg.org/multipage/rendering.html#widgets</a>
<a href="https://drafts.csswg.org/css-ui-4/#appearance-disabling-properties">https://drafts.csswg.org/css-ui-4/#appearance-disabling-properties</a>

* LayoutTests/TestExpectations:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSProperty.h:
* Source/WebCore/css/process-css-properties.py:
(StylePropertyCodeGenProperties):
(GenerateCSSPropertyNames):
* Source/WebCore/dom/Element.h:
(WebCore::Element::isDevolvable const):
* Source/WebCore/html/HTMLButtonElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::isDevolvable const):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLMeterElement.h:
* Source/WebCore/html/HTMLProgressElement.h:
* Source/WebCore/html/HTMLSelectElement.h:
* Source/WebCore/html/HTMLTextAreaElement.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::autoAppearanceForElement const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::RenderStyle):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::isNativeAppearanceDisabled):
(WebCore::RenderStyle::setIsNativeAppearanceDisabled):
(WebCore::RenderStyle::NonInheritedFlags::operator== const):
(WebCore::RenderStyle::NonInheritedFlags::copyNonInheritedFrom):
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyCustomProperty):
(WebCore::Style::Builder::applyCascadeProperty):
(WebCore::Style::Builder::applyRollbackCascadeProperty):
(WebCore::Style::Builder::applyProperty):
* Source/WebCore/style/StyleBuilder.h:
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:
(JSONCSSPropertiesChecker.check_codegen_properties):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/482584f17a16511d68223b83469e1dbf88085e2b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14722 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15030 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15381 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16470 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13883 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14859 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17550 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15128 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16528 "9 flakes 156 failures") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14902 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15403 "2 new passes 4 flakes 14 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12497 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17205 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/14824 "Passed tests") | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12687 "5 flakes 7 failures") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13279 "14 flakes 7 failures") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20273 "2 flakes 120 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13761 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13447 "14 flakes 1 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16687 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13991 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11841 "Exiting early after 60 failures. 42718 tests run. 60 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13290 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17628 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13835 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/style/StyleBuilder.cpp:
```diff
@@ -350,6 +350,9 @@ void Builder::applyProperty(CSSPropertyID id, CSSValue& value, SelectorChecker::
 #endif
 
+    if (cascadeLevel == CascadeLevel::Author && CSSProperty::disablesNativeAppearance(id))
+        style.setIsNativeAppearanceDisabled(true);
```
Can we remove `RenderTheme::isControlStyled` now that we have this?

obyknovenius replied:
I'll have a look.

pxlcoder commented on Source/WebCore/dom/Element.h:
```diff
@@ -544,6 +544,8 @@ class Element : public ContainerNode {
     virtual bool isSliderThumbElement() const { return false; }
     virtual bool isHTMLTablePartElement() const { return false; }
 
+    virtual bool isDevolvable() const { return false; }
```
Maybe `isDevolvableWidget` to match the spec?

obyknovenius replied:
:+1:

pxlcoder commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -1886,6 +1886,25 @@ bool HTMLInputElement::isEmptyValue() const
 
+bool HTMLInputElement::isDevolvable() const
+{
+    return m_inputType->isTextType()
```
I think you can just return `true` here.

From https://html.spec.whatwg.org/multipage/rendering.html#widgets:
```
The following elements can have a native appearance for the purpose of the CSS appearance property.

- button
- input
- meter
- progress
- select
- textarea
```

obyknovenius replied:
It's a little bit more complicated. For example, `<input>` whose `type` is checkbox or radio group is non-devolvable:
https://html.spec.whatwg.org/multipage/rendering.html#the-input-element-as-a-checkbox-and-radio-button-widgets

obyknovenius committed:
Disable native appearance of devolvable widgets
https://bugs.webkit.org/show_bug.cgi?id=243899
rdar://98899342

Reviewed by NOBODY (OOPS!).

All HTML widgets can be divided into two groups: devolvable and
non-devolvable. When certain CSS properties are applied to a devolvable
widget by author styles, the widget must be rendered in its primitive
appearance. Non-devolvable widgets remain native regardless of any such
properties.

Properties that disable native appearance of devolvable widgets are now
marked with `disables-native-appearance` key in `CSSProperites.json`.

https://html.spec.whatwg.org/multipage/rendering.html#widgets
https://drafts.csswg.org/css-ui-4/#appearance-disabling-properties

* LayoutTests/TestExpectations:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSProperty.h:
* Source/WebCore/css/process-css-properties.py:
(StylePropertyCodeGenProperties):
(GenerateCSSPropertyNames):
* Source/WebCore/dom/Element.h:
(WebCore::Element::isDevolvable const):
* Source/WebCore/html/HTMLButtonElement.h:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::isDevolvable const):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLMeterElement.h:
* Source/WebCore/html/HTMLProgressElement.h:
* Source/WebCore/html/HTMLSelectElement.h:
* Source/WebCore/html/HTMLTextAreaElement.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::autoAppearanceForElement const):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::RenderStyle):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::isNativeAppearanceDisabled):
(WebCore::RenderStyle::setIsNativeAppearanceDisabled):
(WebCore::RenderStyle::NonInheritedFlags::operator== const):
(WebCore::RenderStyle::NonInheritedFlags::copyNonInheritedFrom):
* Source/WebCore/style/StyleBuilder.cpp:
(WebCore::Style::Builder::applyCustomProperty):
(WebCore::Style::Builder::applyCascadeProperty):
(WebCore::Style::Builder::applyRollbackCascadeProperty):
(WebCore::Style::Builder::applyProperty):
* Source/WebCore/style/StyleBuilder.h:
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:
(JSONCSSPropertiesChecker.check_codegen_properties):
